### PR TITLE
fastjet/internal/delaunay: handle duplicate points

### PR DESCRIPTION
**Pull request by Bastiantheone**
This PR handles duplicate points. When a duplicate point is inserted it only updates the nearest neighbor of itself and the duplicate point. When it is removed nothing happens.

### CODE REVIEW

File: fastjet/internal/delaunay/delaunay.go [hunk 1]
```diff
@@ -162,6 +169,36 @@ func (d *Delaunay) insertInside(p *Point, t *Triangle) []*Point {
 // insertOnEdge inserts a point on an Edge between two triangles. It returns the points
 // whose nearest neighbor changed.
 func (d *Delaunay) insertOnEdge(p *Point, t *Triangle) []*Point {
+	// Check if p is a duplicate
+	switch {
+	case p.Equals(t.A):
+		if p.id == t.A.id {
+			panic(fmt.Errorf("Point %v was previously inserted", p))
```
Comment (sbinet) [inline fastjet/internal/delaunay/delaunay.go]: please prepend a `"delaunay: ` string to the panic message.

File: fastjet/internal/delaunay/delaunay.go [hunk 2]
```diff
@@ -162,6 +169,36 @@ func (d *Delaunay) insertInside(p *Point, t *Triangle) []*Point {
...
+		}
+		p.nearest = t.A
+		p.dist2 = 0
+		t.A.nearest = p
+		t.A.dist2 = 0
+		return []*Point{p, t.A}
+	case p.Equals(t.B):
+		if p.id == t.B.id {
+			panic(fmt.Errorf("Point %v was previously inserted", p))
```
Comment (sbinet) [inline fastjet/internal/delaunay/delaunay.go]: please prepend a `"delaunay: ` string to the panic message.

File: fastjet/internal/delaunay/delaunay.go [hunk 3]
```diff
@@ -162,6 +169,36 @@ func (d *Delaunay) insertInside(p *Point, t *Triangle) []*Point {
...
+		}
+		p.nearest = t.B
+		p.dist2 = 0
+		t.B.nearest = p
+		t.B.dist2 = 0
+		return []*Point{p, t.B}
+	case p.Equals(t.C):
+		if p.id == t.C.id {
+			panic(fmt.Errorf("Point %v was previously inserted", p))
```
Comment (sbinet) [inline fastjet/internal/delaunay/delaunay.go]: please prepend a `"delaunay: ` string to the panic message.

File: fastjet/internal/delaunay/delaunay.go [hunk 4]
```diff
@@ -162,6 +169,36 @@ func (d *Delaunay) insertInside(p *Point, t *Triangle) []*Point {
 // insertOnEdge inserts a point on an Edge between two triangles. It returns the points
 // whose nearest neighbor changed.
 func (d *Delaunay) insertOnEdge(p *Point, t *Triangle) []*Point {
+	// Check if p is a duplicate
+	switch {
```
Comment (sbinet) [inline fastjet/internal/delaunay/delaunay.go]: can `t` be made of 2 or even 3 duplicates?
e.g.: can `p.Equals(t.A)` and `p.Equals(t.B)` be both true?
presumably a `Triangle` can't be made like so, but... as we now have some degeneracy...

Comment (Bastiantheone) [reply fastjet/internal/delaunay/delaunay.go]: No, because NewTriangle panics when three points are colinear

Comment (sbinet) [reply fastjet/internal/delaunay/delaunay.go]: good :)

File: fastjet/internal/delaunay/delaunay.go (hunk 1 above)
Comment (Bastiantheone) [reply fastjet/internal/delaunay/delaunay.go]: Done

File: fastjet/internal/delaunay/delaunay.go (hunk 2 above)
Comment (Bastiantheone) [reply fastjet/internal/delaunay/delaunay.go]: Done

File: fastjet/internal/delaunay/delaunay.go (hunk 3 above)
Comment (Bastiantheone) [reply fastjet/internal/delaunay/delaunay.go]: Done

File: fastjet/internal/delaunay/delaunay.go (hunk 4 above)
Comment (Bastiantheone) [reply fastjet/internal/delaunay/delaunay.go]: Should I change it to let triangles be composed of duplicates?

Comment (sbinet) [reply fastjet/internal/delaunay/delaunay.go]: no, I believe it's fine as is.
we'll cross that bridge when/if we need to.

**Comment by Bastiantheone** [comment]
Done

**Pull request merged**